diesel/src/deserialize.rs

```diff
@@ -255,6 +256,153 @@ where
+/// struct User {
+///     id: i32,
+///     name: String,
+/// }
+///
+/// # fn main() {
+/// #     run_test();
+/// # }
+/// #
+/// # fn run_test() -> QueryResult<()> {
+/// #     use schema::users::dsl::*;
+/// #     let connection = establish_connection();
+/// let first_user = users.first(&connection)?;
+/// let expected = User { id: 1, name: "Sean".into() };
+/// assert_eq!(expected, first_user);
+/// #     Ok(())
+/// # }
+/// ```
+///
+/// If we want to do additional work during deserialization, we can use
+/// `deserialize_as` to use a different implementation.
+///
+/// ```rust
+/// # include!("doctest_setup.rs");
+/// #
+/// # use schema::users;
+/// # use diesel::backend::{self, Backend};
+/// # use diesel::deserialize::Queryable;
+/// #
+/// struct LowercaseString(String);
+///
+/// impl Into<String> for LowercaseString {
+///     fn into(self) -> String {
+///         self.0
+///     }
+/// }
+///
+/// impl<DB, ST> Queryable<ST, DB> for LowercaseString
+/// where
+///     DB: Backend,
+///     String: Queryable<ST, DB>,
+/// {
+///     type Row = <String as Queryable<ST, DB>>::Row;
+///
+///     fn build(row: Self::Row) -> Self {
+///         LowercaseString(String::build(row).to_lowercase())
+///     }
+/// }
+///
+/// #[derive(Queryable, PartialEq, Debug)]
+/// struct User {
+///     id: i32,
+///     #[diesel(deserialize_as = "LowercaseString")]
+///     name: String,
+/// }
+///
+/// # fn main() {
+/// #     run_test();
+/// # }
+/// #
+/// # fn run_test() -> QueryResult<()> {
+/// #     use schema::users::dsl::*;
+/// #     let connection = establish_connection();
+/// let first_user = users.first(&connection)?;
+/// let expected = User { id: 1, name: "sean".into() };
+/// assert_eq!(expected, first_user);
+/// #     Ok(())
+/// # }
+/// ```
+///
+/// Alternatively, we can implement the trait for our struct manually.
+///
+/// ```rust
+/// # include!("doctest_setup.rs");
+/// #
+/// use schema::users;
+/// use diesel::deserialize::Queryable;
+///
+/// # /*
+/// type DB = diesel::sqlite::Sqlite;
+/// # */
+///
+/// #[derive(PartialEq, Debug)]
+/// struct User {
+///     id: i32,
+///     name: String,
+/// }
+///
+/// impl Queryable<users::SqlType, DB> for User {
+///     type Row = (i32, String);
+///
+///     fn build(row: Self::Row) -> Self {
+///         User {
+///             id: row.0,
+///             name: row.1.to_lowercase(),
+///         }
+///     }
+/// }
+///
+/// # fn main() {
+/// #     run_test();
+/// # }
+/// #
+/// # fn run_test() -> QueryResult<()> {
+/// #     use schema::users::dsl::*;
+/// #     let connection = establish_connection();
+/// let first_user = users.first(&connection)?;
+/// let expected = User { id: 1, name: "sean".into() };
+/// assert_eq!(expected, first_user);
+/// #     Ok(())
+/// # }
+/// ```
+pub trait TableQueryable {
+    /// The Rust type you'd like to map from.
+    ///
+    /// This is typically a tuple of all of your struct's fields.
+    type Columns: Expression;
```

Comment:
Should probably be called `SelectExpression` or something, because it does not necessarily only need to contain a list of columns. For a manual implementation this could be anything in theory.

(Same for the corresponding method)

Reply:
Why not just `Expression`? Would `SelectExpression` be confused with trait `SelectableExpression`?

Reply:
BTW, I think any expression could be called "column" in `SELECT` statement, at least in [postgres](https://www.postgresql.org/docs/current/sql-select.html)
> Just as in a table, every output column of a SELECT has a name.
> An output column's name can be used to refer to the column's value in ORDER BY and GROUP BY clauses
> In more complex cases a function or type name may be used, or the system may fall back on a generated name such as ?column?.

Just like every item returned from a query called rows.

Reply:
Sure. But just curious, I've done a search `/select.*expression/` and it only shows `SelectableExpression` which I think have a different meaning. Where is "other place"?
I never consist on naming, but I just want some name with consistence.

Reply:
In my understanding `SelectableExpression` says that something can be part of the select clause, where `SelectExpression` says it will be part of the select clause. So both refer to the same clause and both are expressions, while one states a possibility, while the other states a fact. 
That said: I'm also no native speaker, so this could be my non-prefect English, so if someone has other interpretations here I'm happy to be corrected.